File: Assets/Bundles/Shared/Resources/TS/VoxelWorld/World.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Game = require("Shared/TS/Game").Game
local GetItemMeta = require("Shared/TS/Item/ItemDefinitions").GetItemMeta
local Network = require("Shared/TS/Network").Network
local Signal = require("Shared/TS/Util/Signal").Signal
local Block = require("Shared/TS/VoxelWorld/Block").Block
local VoxelDataAPI = require("Shared/TS/VoxelWorld/VoxelData/VoxelDataAPI").VoxelDataAPI
local World
do
	World = setmetatable({}, {
		__tostring = function()
			return "World"
		end,
	})
	World.__index = World
	function World.new(...)
		local self = setmetatable({}, World)
		return self:constructor(...) or self
	end
	function World:constructor(voxelWorld)
		self.voxelWorld = voxelWorld
		self.OnVoxelPlaced = Signal.new()
		voxelWorld:OnVoxelPlaced(function(voxel, x, y, z)
			local vec = Vector3.new(x, y, z)
			VoxelDataAPI:ClearVoxelData(vec)
			voxel = VoxelWorld:VoxelDataToBlockId(voxel)
			self.OnVoxelPlaced:Fire(vec, voxel)
		end)
	end
	function World:GetRawVoxelDataAt(pos)
		return self.voxelWorld:ReadVoxelAt(pos)
	end
	function World:GetBlockAt(pos)
		return Block.new(self.voxelWorld:ReadVoxelAt(pos), self)
	end
	function World:GetBlockBelowMeta(pos)
		local _fn = self
		local _pos = pos
		local _vector3 = Vector3.new(0, -0.1, 0)
		local _result = _fn:GetBlockAt(_pos + _vector3)
		if _result ~= nil then
			_result = _result.itemMeta
			if _result ~= nil then
				_result = _result.block
			end
		end
		return _result
	end
	function World:PlaceBlock(pos, itemType, config)
		local itemMeta = GetItemMeta(itemType)
		if not itemMeta.block then
			return nil
		end
		local blockId = itemMeta.block.blockId
		self:PlaceBlockById(pos, blockId, config)
	end
	function World:PlaceBlockById(pos, blockId, config)
		local _fn = self.voxelWorld
		local _exp = pos
		local _exp_1 = blockId
		local _result = config
		if _result ~= nil then
			_result = _result.priority
		end
		local _condition = _result
		if _condition == nil then
			_condition = true
		end
		_fn:WriteVoxelAt(_exp, _exp_1, _condition)
		if RunCore:IsServer() then
			local _fn_1 = Network.ServerToClient.BlockPlace.Server
			local _exp_2 = pos
			local _exp_3 = blockId
			local _result_1 = config
			if _result_1 ~= nil then
				_result_1 = _result_1.placedByEntityId
			end
			_fn_1:FireAllClients(_exp_2, _exp_3, _result_1)
		else
			local _result_1 = config
			if _result_1 ~= nil then
				_result_1 = _result_1.placedByEntityId
			end
			local _result_2 = Game.LocalPlayer.Character
			if _result_2 ~= nil then
				_result_2 = _result_2.id
			end
			if _result_1 == _result_2 then
				-- Client predicted block place event
				local clientSignals = TS.Promise.new(function(resolve)
					resolve(require("Client/TS/ClientSignals"))
				end):expect().ClientSignals
				local BlockPlaceClientSignal = TS.Promise.new(function(resolve)
					resolve(require("Client/TS/Signals/BlockPlaceClientSignal"))
				end):expect().BlockPlaceClientSignal
				local block = Block.new(blockId, self)
				clientSignals.BlockPlace:Fire(BlockPlaceClientSignal.new(pos, block, Game.LocalPlayer.Character))
			end
		end
	end
	function World:LoadWorldFromVoxelBinaryFile(binaryFile, blockDefines)
		self.voxelWorld:LoadWorldFromVoxelBinaryFile(binaryFile, blockDefines)
	end
	function World:LoadEmptyWorld(blockDefines, cubeMapPath)
		self.voxelWorld:LoadEmptyWorld(blockDefines, cubeMapPath)
	end
	function World:RaycastVoxel(pos, direction, maxDistance)
		return self.voxelWorld:RaycastVoxel(pos, direction, maxDistance)
	end
	function World:GetBlockDefinition(blockId)
		return self.voxelWorld.blocks:GetBlockDefinitionFromIndex(blockId)
	end
	function World:GetBlockAverageColor(blockId)
		local _result = self:GetBlockDefinition(blockId)
		if _result ~= nil then
			_result = _result.averageColor:GetValue(0)
		end
		return _result
	end
	World.SKYBOX = "Shared/Resources/Skybox/BrightSky/bright_sky_2.png"
end
return {
	World = World,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/VoxelWorld/World.ts
import Object from "@easy-games/unity-object-utils";
import { Game } from "Shared/Game";
import { ItemType } from "Shared/Item/ItemType";
import { Network } from "Shared/Network";
import { Signal } from "Shared/Util/Signal";
import { BlockMeta } from "../Item/ItemMeta";
import { ItemUtil } from "../Item/ItemUtil";
import { Block } from "./Block";
import { BlockDataAPI } from "./BlockData/BlockDataAPI";

export interface PlaceBlockConfig {
	placedByEntityId?: number;
	priority?: boolean;
	blockData?: {
		[key: string]: unknown;
	};
}

export class World {
	public static SKYBOX = "Shared/Resources/Skybox/BrightSky/bright_sky_2.png";

	public OnVoxelPlaced = new Signal<[pos: Vector3, voxel: number]>();
	public OnFinishedLoading = new Signal<void>();
	public OnFinishedReplicatingChunksFromServer = new Signal<void>();
	private finishedLoading = false;
	private finishedReplicatingChunksFromServer = false;

	constructor(public readonly voxelWorld: VoxelWorld) {
		voxelWorld.OnVoxelPlaced((voxel, x, y, z) => {
			const vec = new Vector3(x, y, z);
			BlockDataAPI.ClearBlockData(vec);
			voxel = VoxelWorld.VoxelDataToBlockId(voxel);
			this.OnVoxelPlaced.Fire(vec, voxel);
		});

		if (!voxelWorld.finishedLoading) {
			voxelWorld.OnFinishedLoading(() => {
				this.finishedLoading = true;
				this.OnFinishedLoading.Fire();
			});
		} else {
			this.finishedLoading = true;
			this.OnFinishedLoading.Fire();
		}

		if (!voxelWorld.finishedReplicatingChunksFromServer) {
			voxelWorld.OnFinishedReplicatingChunksFromServer(() => {
				this.finishedReplicatingChunksFromServer = true;
				this.OnFinishedReplicatingChunksFromServer.Fire();
			});
		} else {
			this.finishedReplicatingChunksFromServer = true;
			this.OnFinishedReplicatingChunksFromServer.Fire();
		}
	}

	public IsFinishedLoading(): boolean {
		return this.finishedLoading;
	}

	public async WaitForFinishedLoading(): Promise<void> {
		if (this.finishedLoading) return;

		return new Promise<void>((resolve) => {
			this.OnFinishedLoading.Wait();
			resolve();
		});
	}

	public IsFinishedReplicatingChunksFromServer(): boolean {
		return this.finishedReplicatingChunksFromServer;
	}

	public async WaitForFinishedReplicatingChunksFromServer(): Promise<void> {
		if (this.finishedReplicatingChunksFromServer) return;

		return new Promise<void>((resolve) => {
			this.OnFinishedReplicatingChunksFromServer.Wait();
			resolve();
		});
	}

	/**
	 *
	 * @param pos
	 * @returns Raw voxel data.
	 * @deprecated Use `GetVoxelAt()` instead.
	 */
	public GetRawVoxelDataAt(pos: Vector3): number {
		return this.voxelWorld.ReadVoxelAt(pos);
	}

	/**
	 * A more convenient version of ReadVoxelAt.
	 * @param pos
	 * @returns VoxelBlock at position.
	 */
	public GetBlockAt(pos: Vector3): Block {
		return new Block(this.voxelWorld.ReadVoxelAt(pos), this);
	}

	/**
	 * A way to find block data below a target. Used to know what a character is standing on
	 * @param pos
	 * @returns BlockMeta under the position.
	 */
	public GetBlockBelowMeta(pos: Vector3): BlockMeta | undefined {
		return this.GetBlockAt(pos.add(new Vector3(0, -0.5, 0)))?.itemMeta?.block;
	}

	public PlaceBlock(pos: Vector3, itemType: ItemType, config?: PlaceBlockConfig): void {
		const itemMeta = ItemUtil.GetItemMeta(itemType);
		if (!itemMeta.block) return;

		const blockId = itemMeta.block.blockId;
		this.PlaceBlockById(pos, blockId, config);
	}

	public PlaceBlockById(pos: Vector3, blockId: number, config?: PlaceBlockConfig): void {
		this.voxelWorld.WriteVoxelAt(pos, blockId, config?.priority ?? true);
		if (config?.blockData) {
			for (const key of Object.keys(config.blockData)) {
				BlockDataAPI.SetBlockData(pos, key as string, config.blockData[key]);
			}
		}

		if (RunCore.IsServer()) {
			Network.ServerToClient.BlockPlace.Server.FireAllClients(pos, blockId, config?.placedByEntityId);
		} else {
			if (config?.placedByEntityId === Game.LocalPlayer.Character?.id) {
				// Client predicted block place event
				const clientSignals = import("Client/ClientSignals").expect().ClientSignals;
				const BlockPlaceClientSignal = import("Client/Signals/BlockPlaceClientSignal").expect()
					.BlockPlaceClientSignal;

				const block = new Block(blockId, this);
				clientSignals.BlockPlace.Fire(new BlockPlaceClientSignal(pos, block, Game.LocalPlayer.Character));
			}
		}
	}

	public LoadWorldFromVoxelBinaryFile(binaryFile: VoxelBinaryFile, blockDefines: TextAsset): void {
		this.voxelWorld.LoadWorldFromVoxelBinaryFile(binaryFile, blockDefines);
	}

	public LoadEmptyWorld(blockDefines: TextAsset, cubeMapPath: string): void {
		this.voxelWorld.LoadEmptyWorld(blockDefines, cubeMapPath);
	}

	public RaycastVoxel(pos: Vector3, direction: Vector3, maxDistance: number): VoxelRaycastResult {
		return this.voxelWorld.RaycastVoxel(pos, direction, maxDistance);
	}

	public GetBlockDefinition(blockId: number): BlockDefinition | undefined {
		return this.voxelWorld.blocks.GetBlockDefinitionFromIndex(blockId);
	}

	public GetBlockAverageColor(blockId: number): Color | undefined {
		return this.GetBlockDefinition(blockId)?.averageColor.GetValue(0);
	}
}
